packages/context-editor/src/plugins/attributePanel.ts

```diff
@@ -1,19 +1,195 @@
+				// 		])
+				// 	);
+				// 	console.log({ pos, node, event });
+				// 	setPanelPosition({
+				// 		...panelPosition,
+				// 		isOpen: !panelPosition.isOpen,
+				// 		isLink,
+				// 		pos: range?.from || pos,
+				// 		node,
+				// 	});
+
+				// 	return true;
+				// },
 			},
-			apply: (tr, prev) => tr.getMeta(attributePanelKey) || prev,
-		},
-	});
+			// view: () => {
+			// 	return {
+			// 		update: (editorView, prev) => {
+			// 			const pos = editorView.state.selection.$from.pos;
+			// 			const prevPos = prev.selection.$from.pos;
+
+			// 			if (pos !== undefined && prevPos !== pos) {
+			// 				console.log({
+			// 					msg: "updating panel pos",
+			// 					pos,
+			// 					prevPos,
+			// 				});
+			// const isLink = Boolean(
+			// 	baseSchema.marks.link.isInSet([
+			// 		...(editorView.state.storedMarks || []),
+			// 		...(editorView.state.selection.$from.marks() || []),
+			// 	])
+			// );
+			// 				const coords = editorView.coordsAtPos(pos, 1);
+			// 				let isOpen = panelPosition.isOpen;
+			// 				if (isLink) {
+			// 					isOpen = true;
+			// 				}
+			// 				const node =
+			// 					editorView.state.selection.$from.nodeAfter || panelPosition.node;
+			// 				if (node?.isText && node.marks?.length === 0) {
+			// 					isOpen = false;
+			// 				}
+			// 				setPanelPosition({
+			// 					...panelPosition,
+			// 					selection: editorView.state.selection,
+			// 					isOpen,
+			// 					isLink,
+			// 					pos,
+			// 				});
+			// 			}
+			// 		},
+			// 	};
+			// },
+		}),
+		new Plugin({
+			view: pluginViewFactory({
+				component: () => AttributePanel(),
+				root: (viewDom) =>
+					document.getElementById("context-editor-container") as HTMLElement,
+			}),
+		}),
+	];
 };
+
+function getUpdatedDecorations(
+	tr: Transaction,
+	pluginState: AttributePanelState,
+	prevEditorState: EditorState,
+	editorState: EditorState,
+	getToggleWidget: WidgetDecorationFactory
+) {
+	let prevDecorationSet = pluginState.decorations;
+	const decorations: Decoration[] = [];
+	if (!prevDecorationSet) {
+		console.log("generating initial decoration set");
+
+		editorState.doc.descendants((node, pos) => {
+			if (node.type.isBlock) {
+				decorations.push(getToggleWidget(pos, { node }));
+			}
+			const isInline = !node.type.isBlock;
+			const hasMarks = !!node.marks.length;
+			const isMath = node.type.name === "math_inline";
+			if (isInline && (hasMarks || isMath)) {
+				/* If it's an inline node with marks OR is inline math */
+				decorations.push(getToggleWidget(pos, { node }));
+			}
+		});
+		return DecorationSet.create(editorState.doc, decorations);
+	}
+	let mappedDecorations = prevDecorationSet.map(tr.mapping, editorState.doc);
+
+	for (const stepMap of tr.mapping.maps) {
+		stepMap.forEach((oldStart, oldEnd, newStart, newEnd) => {
+			mappedDecorations = mappedDecorations.remove(mappedDecorations.find(newStart, newEnd));
+			editorState.doc.nodesBetween(newStart, newEnd, (node, pos, parent, index) => {
```

Comment:
I haven't tested this, so if you do use it, I'd make sure this is using the right combo of old/new positions and the respected editorstate vs previouseditorstate. Also there's a chance these are just the totally wrong functions to use 😓